engine/src/lua_setup.cpp

```diff
@@ -0,0 +1,343 @@
+  std::string package_path = lua["package"]["path"];
+  for (const std::string& p : paths) {
+    package_path += ";" + p + "/?.lua";
+    package_path += ";" + p + "/?/init.lua";
+  }
+  lua["package"]["path"] = package_path;
+}
+
+/**
+ * Add Lua package paths so that bundled Lua libaries can be found.
+ */
+void register_package_path(sol::state_view& lua, const LuaOptions& opt) {
+  // Setup lua path:
+  std::vector<std::string> lua_path{};
+  if (!opt.no_system_lua) {
+    // FIXME(windows): These paths are linux-specific.
+    lua_path = {
+        "/usr/local/lib/cloe/lua",
+        "/usr/lib/cloe/lua",
+    };
+  }
+  std::string lua_paths = opt.environment->get_or(CLOE_LUA_PATH, "");
+  for (auto&& p : utility::split_string(std::move(lua_paths), ":")) {
+    if (contains(lua_path, p)) {
+      continue;
+    }
+    lua_path.emplace_back(std::move(p));
+  }
+  for (const auto& p : opt.lua_paths) {
+    if (contains(lua_path, p)) {
+      continue;
+    }
+    lua_path.emplace_back(p);
+  }
+
+  configure_package_path(lua, lua_path);
+}
+
+/**
+ * Load "cloe-engine" library into Lua.
+ *
+ * This is then available via:
+ *
+ *     require("cloe-engine")
+ *
+ * Any changes you make here should be documented in the Lua meta files.
+ *
+ *     engine/lua/cloe-engine/init.lua
+ */
+void register_cloe_engine(sol::state_view& lua, Stack& stack) {
+  sol::table tbl = lua.create_table();
+
+  // Initial input will be processed at simulation start.
+  tbl["initial_input"] = lua.create_table();
+  tbl["initial_input"]["triggers"] = lua.create_table();
+  tbl["initial_input"]["triggers_processed"] = 0;
+  tbl["initial_input"]["signal_aliases"] = lua.create_table();
+  tbl["initial_input"]["signal_requires"] = lua.create_table();
+
+  // Plugin access will be made available by Coordinator.
+  tbl["plugins"] = lua.create_table();
+
+  // Simulation state will be extended in simulation.
+  // clang-format off
+  tbl["state"] = lua.create_table();
+  tbl["state"]["report"] = lua.create_table();
+  tbl["state"]["stack"] = std::ref(stack);
+  tbl["state"]["config"] = fable::into_sol_object(lua, stack.active_config());
+  tbl["state"]["scheduler"] = sol::lua_nil;
+  tbl["state"]["current_script_file"] = sol::lua_nil;
+  tbl["state"]["current_script_dir"] = sol::lua_nil;
+  tbl["state"]["scripts_loaded"] = lua.create_table();
+  tbl["state"]["features"] = lua.create_table_with(
+    // Version compatibility:
+    "cloe-0.18.0", true,
+    "cloe-0.18", true,
+    "cloe-0.19.0", true,
+    "cloe-0.19", true,
+    "cloe-0.20.0", true,
+    "cloe-0.20", true,
+    "cloe-0.21.0", true, // nightly
+    "cloe-0.21", true,   // nightly
+
+    // Stackfile versions support:
+    "cloe-stackfile", true,
+    "cloe-stackfile-4", true,
+    "cloe-stackfile-4.0", true,
+    "cloe-stackfile-4.1", true,
+
+    // Server enabled:
+    "cloe-server", CLOE_ENGINE_WITH_SERVER != 0,
+    "cloe-lrdb", CLOE_ENGINE_WITH_LRDB != 0
+  );
+  // clang-format on
+
+#if 0
```

Comment:
This is the API I want but I was getting UB / segfaults I think with it used. I'll see if I can fix again, otherwise I'll move it to a separate commit & PR.

Reply:
TODO:
- [ ] Fix dead code in "lua_setup.cpp"

Reply:
This has now been fixed.